lib/Sema/TypeCheckStorage.cpp

```diff
@@ -705,22 +727,43 @@ static Expr *buildStorageReference(AccessorDecl *accessor,
       firstWrapperIdx = 1;
 
     // Perform accesses to the wrappedValues along the composition chain.
-    for (unsigned i : range(firstWrapperIdx, lastWrapperIdx)) {
-      auto wrapperInfo = var->getAttachedPropertyWrapperTypeInfo(i);
-      auto wrappedValue = wrapperInfo.valueVar;
-
-      // Check for availability of wrappedValue.
-      if (accessor->getAccessorKind() == AccessorKind::Get ||
-          accessor->getAccessorKind() == AccessorKind::Read) {
-        if (auto *attr = wrappedValue->getAttrs().getUnavailable(ctx)) {
-          diagnoseExplicitUnavailability(
-              wrappedValue,
-              var->getAttachedPropertyWrappers()[i]->getRangeWithAt(),
-              var->getDeclContext(), nullptr);
+    if (firstWrapperIdx < lastWrapperIdx) {
+      if (auto lvalueness = getPropertyWrapperLValueness(var)) {
+
+        // Figure out if the outermost wrapper instance should be an l-value
+        bool isLValueForGet = lvalueness->isLValueForGetAccess[firstWrapperIdx];
+        bool isLValueForSet = lvalueness->isLValueForSetAccess[firstWrapperIdx];
+        isMemberLValue = (isLValueForGet && isUsedForGetAccess) ||
+                         (isLValueForSet && isUsedForSetAccess);
+
+        for (unsigned i : range(firstWrapperIdx, lastWrapperIdx)) {
+          auto wrapperInfo = var->getAttachedPropertyWrapperTypeInfo(i);
+          auto wrappedValue = wrapperInfo.valueVar;
+
+          // Figure out if the wrappedValue accesses should be l-values
+          bool isWrapperRefLValue = isLValue;
+          if (i < lastWrapperIdx - 1) {
+            bool isLValueForGet = lvalueness->isLValueForGetAccess[i+1];
+            bool isLValueForSet = lvalueness->isLValueForSetAccess[i+1];
```

Comment:
Why are these accessing `i + 1` instead of `i`?

Reply:
Given Swift code like `@A @B @C var member: Int`, we're building an expression like `self._member.wrappedValue.wrappedValue.wrappedValue`.

```
@A @B @C var member: Int

self    _member     wrappedValue    wrappedValue    wrappedValue
        (of type A) (of type B)     (of type C)     (of type Int)

       |                                        |  |            |
       +----------------------------------------+  +------------+
       Based on PropertyWrapperLValueness          Based on
                (lvalueness)                       isUsedForSetAccess
                                                   argument
       |         | |                                            |
       +---------+ +--------------------------------------------+
       Tracked by       Tracked by
       isMemberLValue   underlyingVars[i].second
```

Consequently:

  - isMemberLValue <- lvalueness[0]
  - underlyingVars[0].second <- lvalueness[1]
  - underlyingVars[1].second <- lvalueness[2]
  - underlyingVars[2].second <- isUsedForSetAccess